File: apps/android/core/model/src/main/kotlin/com/sigmundgranaas/turbo/expressive/domain/SavedPath.kt
package com.sigmundgranaas.turbo.expressive.domain

import com.sigmundgranaas.turbo.expressive.core.geo.GeoPath
import com.sigmundgranaas.turbo.expressive.core.geo.PhaseSplit

/** A named, persisted [GeoPath] — a recorded track or saved route, optionally
 *  tagged with the [activityKind] it represents (hiking, skiing, fishing…).
 *
 *  When the track came from following a planned route (D1), [plannedRoute] keeps
 *  the guide geometry it was walked against and [phaseSplits] the checkpoint splits
 *  recorded along the way, so the saved artifact can redraw the guide + splits later.
 *  Both are empty/null for plain recordings. */
data class SavedPath(
    val id: String,
    val name: String,
    val path: GeoPath,
    val activityKind: ActivityKindId? = null,
    val plannedRoute: List<LatLng>? = null,
    val phaseSplits: List<PhaseSplit> = emptyList(),
    /** Display colour as "#RRGGBB", null = the default track colour. Synced
     *  (`TrackMetadataDto.colorHex`) so a colour picked on any client renders
     *  everywhere. */
    val colorHex: String? = null,
    /** Icon key chosen for the track (web edits this today); pass-through synced. */
    val iconKey: String? = null,
    /** Line-style key (`solid`/`dotted`/`dashed`/`dash_dot`); pass-through synced.
     *  Android renders tracks as 3D tubes, which draw solid regardless — the key is
     *  kept so a style picked on another client survives an Android edit. */
    val lineStyleKey: String? = null,
)

File: apps/ios/Sources/CoreModel/SavedPath.swift
import Foundation

/// A named, persisted ``GeoPath`` — a recorded track or saved route, optionally
/// tagged with the activity it represents. Mirrors `domain.SavedPath`.
public struct SavedPath: Identifiable, Hashable, Sendable {
    public let id: String
    public let name: String
    public let path: GeoPath
    public let activityKind: ActivityKindId?
    /// When this track came from following a planned route (D1): the guide it was walked
    /// against, so the saved artifact can redraw it. Nil for plain recordings.
    public let plannedRoute: [LatLng]?
    /// Checkpoint splits recorded while following (D1); empty for plain recordings.
    public let phaseSplits: [PhaseSplit]

    public init(id: String, name: String, path: GeoPath, activityKind: ActivityKindId? = nil,
                plannedRoute: [LatLng]? = nil, phaseSplits: [PhaseSplit] = []) {
        self.id = id
        self.name = name
        self.path = path
        self.activityKind = activityKind
        self.plannedRoute = plannedRoute
        self.phaseSplits = phaseSplits
    }

    /// `recordedAtEpochMs` as a `Date`, when present.
    public var recordedAt: Date? {
        path.recordedAtEpochMs.map { Date(timeIntervalSince1970: Double($0) / 1000) }
    }
}
